compiler-core/src/type_/expression.rs

```diff
@@ -18,123 +18,184 @@ use im::hashmap;
+    /// ```
+    pub fn no_externals() -> Implementations {
+        Implementations {
+            gleam: true,
+            erlang: false,
+            javascript: false,
         }
     }
 
-    pub fn javascript() -> SupportedTargets {
-        SupportedTargets {
+    /// For values that only uses a javascript external implementation.
+    ///
+    /// Some examples:
+    ///
+    /// ```gleam
+    /// @external(javascript, "foo", "bar")
+    /// pub fn javascript_only() -> Int
+    /// ```
+    pub fn javascript() -> Implementations {
+        Implementations {
+            gleam: false,
             erlang: false,
             javascript: true,
         }
     }
 
-    pub fn erlang() -> SupportedTargets {
-        SupportedTargets {
+    /// For values that only uses an erlang external implementation.
+    ///
+    /// Some examples:
+    ///
+    /// ```gleam
+    /// @external(erlang, "foo", "bar")
+    /// pub fn erlang_only() -> Int
+    /// ```
+    pub fn erlang() -> Implementations {
+        Implementations {
+            gleam: false,
             erlang: true,
             javascript: false,
         }
     }
 
-    pub fn from_target(target: Target) -> SupportedTargets {
-        match target {
-            Target::Erlang => SupportedTargets::erlang(),
-            Target::JavaScript => SupportedTargets::javascript(),
-        }
-    }
-
-    pub fn intersect(&self, targets: SupportedTargets) -> SupportedTargets {
-        SupportedTargets {
-            erlang: self.erlang && targets.erlang,
-            javascript: self.javascript && targets.javascript,
-        }
-    }
-
-    pub fn merge(&self, targets: SupportedTargets) -> SupportedTargets {
-        SupportedTargets {
-            erlang: self.erlang || targets.erlang,
-            javascript: self.javascript || targets.javascript,
-        }
-    }
-
-    pub fn add(&self, target: Target) -> SupportedTargets {
-        match target {
-            Target::Erlang => SupportedTargets {
-                erlang: true,
-                javascript: self.javascript,
-            },
-            Target::JavaScript => SupportedTargets {
-                erlang: self.erlang,
-                javascript: true,
-            },
+    /// For values that have anexternal implementation for _all_ gleam targets
+    /// but do not have a default Gleam implementation.
+    ///
+    /// Some examples:
+    ///
+    /// ```gleam
+    /// @external(erlang, "foo", "bar")
+    /// @external(javascript, "foo", "bar")
+    /// pub fn foo() -> Int
+    /// ```
+    pub fn all_externals() -> Implementations {
+        Implementations {
+            gleam: false,
+            erlang: true,
+            javascript: true,
         }
     }
 
+    /// Given the implementations of a function update those with taking into
+    /// account the `implementations` of another function (or constant) used
+    /// inside its body.
+    pub fn update_from_use(&mut self, implementations: &Implementations) {
```

Comment:
I removed all other combinators like `merge`/`intersect`/... in favour of just this function that is actually all that is needed by the typer. I think it makes things easier to follow and harder to mess up inadvertently 😁

Reply:
Yes I agree!